### PR TITLE
MBS-8088 impact analysis screenToModule relation via psi

### LINKED ISSUE #292: UI Impact analysis through module pointers
## Problem

Android gradle plugin 3.6 makes namespacedRClass on by default, so transitive resources doesn't merge up through modules graph.
It breaks our approach to map R.id constants with it's modules as id's are no longer statically resolved.

## Proposal

We could add new field for `com.avito.android.screen.Screen`:

```kotlin
interface Screen {
    val rootId: Int
+   val module: String
}
```

`rootId` is still needed for isOpened method

`module` is a module path, like `:lib:foo`, it should point to the closest to app module containing this screen logic

Q: Will at be harder to support?
A: A bit, modules are more or less stable, but we lost isOpened guarantees (isOpened check kept rootId actual)

Q: What if someone renames module?
A: if `module` field point on nonexistent module we could fail build on analysis with descriptive error

Q: What if screen moved to another module?
A: Incorrect module pointers will lead to incorrect ui impact analysis results. Runtime check could be added to match computed rootId with module based on info from module's intermediates

## Future

We are going to left a small(and decreasing) amount of e2e tests in app module, and move component tests to library modules, so this impact analysis approach will be less effective, as libraries inherits standard "per module"
impact analysis.

Why not remove all ui impact analysis code at all?
1. We want to keep our PR times at current level before component tests goes to modules
1. Some of it's logic could be reused, for example to find new and changed tests on PR

### PR DESCRIPTION
**Pull request by dsvoronin**
fixes #292, but with new implementation, just analyzing psi of Screen files to get package modules

### CODE REVIEW

File: subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.avito.impact.util
+
+import java.io.File
+
+object AndroidManifestPackageParser {
+
+    private val manifestPackagePattern = Regex(" package=\"([a-zA-Z0-9._]+)\"")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt]:
```suggestion
    private val manifestPackagePattern = Regex(" package=\"([^"]+)\"")
```

I see no reasons to know about package valid value in this case.

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt [hunk 2]
```diff
@@ -303,22 +315,28 @@ abstract class TestBytecodeAnalyzeAction : WorkAction<TestBytecodeAnalyzeAction.
...
-        val duplicateCheckResult = DuplicateIdChecker.Impl(config.unknownRootId.get()).check(screenToIds)
-        if (duplicateCheckResult.hasDuplicates) {
-            ciLogger.info("There are duplicate id's: ${duplicateCheckResult.ids}")
-        }
+    private fun getScreenToModule(projectModules: Collection<InternalModule>): Collection<ScreenToModulePath> {
+
+        val sourceSets = projectModules.flatMap { it.androidTestConfiguration.sourceSets() }
 
-        return screenToIds.toMap()
+        ciLogger.info("projectModules=$projectModules")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt]: debug ?

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt [hunk 3]
```diff
@@ -303,22 +315,28 @@ abstract class TestBytecodeAnalyzeAction : WorkAction<TestBytecodeAnalyzeAction.
...
+
+        val androidModules = projectModules.filter {
+            it.project.pluginManager.hasPlugin("com.android.library") || it.project.pluginManager.hasPlugin("com.android.application")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt]:
```suggestion
            it.project.isAndroid()
```

We have an extension already.

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+data class ModulePath(val path: String) {
+    init {
+        require(path.matches(validModulePath)) { "$path is not a valid gradle module path" }
+    }
+}
+
+fun ModulePath.toFileName() = path.drop(1).replace(":", "+")
+
+private val validModulePath = Regex("(:[a-zA-Z1-9\\-]+(_[a-zA-Z1-9\\-])?)*")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]:
```suggestion
private val validModulePath = Regex("(:[\w\d\\-]+(_[\w\d\\-])?)*")
```
It's good to use character classes for readability.

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.avito.instrumentation.impact.metadata
+
+data class ModulePath(val path: String) {
+    init {
+        require(path.matches(validModulePath)) { "$path is not a valid gradle module path" }
+    }
+}
+
+fun ModulePath.toFileName() = path.drop(1).replace(":", "+")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: Can paths clash after dropping a caught group?

**Review by eugene-krivobokov — CHANGES_REQUESTED** [review]

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt]: yep, thanks

File: subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt (hunk 1 above)
Comment (dsvoronin) [reply subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt]: i agree

Comment (sergioboi) [reply subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt]: Could we use Xml Parser?

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt (hunk 5 above)
Comment (sergioboi) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: Could we use Gradle API? We may get project dir by a path. Parsing string, dropping ':' looks fragile.
Not a blocker

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt (hunk 4 above)
Comment (sergioboi) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: Yet another great part to delegate to Gradle

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+                                    val importLines = ktFile.importDirectives.map { it.importedFqName.toString() }
+                                    result[fullClassName] = getPackageName(expression.text, importLines)!!
+                                }
+                            }
+                        }
+                }
+        }
+
+        ciLogger.info("Analyzed screen classes in ${System.currentTimeMillis() - startTime}ms")
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt]: debug

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                            val fullClassName = if (packageName.isBlank()) {
+                                className
+                            } else {
+                                "$packageName.$className"
+                            }
+
+                            it.body?.properties?.forEach { property ->
+                                if (property.name == fieldName) {
+
+                                    val expression = property.children.filterIsInstance<KtExpression>().single()
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt]: Could you create a task for assigning via functions and getters?

**Review by sergioboi — APPROVED** [review]

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt (hunk 4 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: As far as I see, it's a custom format similar to the gradle path.

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt (hunk 1 above)
Comment (dsvoronin) [reply subprojects/gradle/impact-shared/src/main/java/com/avito/impact/util/AndroidManifestPackageParser.kt]: > Could we use Xml Parser?

yes, but we don't need deep file structure knowledge. i can optimize it reading line by line

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt (hunk 5 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: i used https://github.com/gradle/gradle/blob/master/subprojects/base-services/src/main/java/org/gradle/util/Path.java , seems like what we need here

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt (hunk 6 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt]: thanks, changed.

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt (hunk 7 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/MetadataParser.kt]: supported it in new commit (kinda, maybe missed some corner cases, will see)

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt (hunk 4 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: used gradle one

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/TestBytecodeAnalyzeAction.kt]: thanks, fixed

File: subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt (hunk 5 above)
Comment (eugene-krivobokov) [reply subprojects/gradle/instrumentation-test-impact-analysis/src/main/kotlin/com/avito/instrumentation/impact/metadata/ModulePath.kt]: It doesn't have validation. But without the user's input it's not needed.
